gr-digital/lib/costas_loop_cc_impl.cc

```diff
@@ -204,6 +224,189 @@ int costas_loop_cc_impl::work(int noutput_items,
+                               gr_vector_void_star& output_items)
+{
+    const StructComplex* iptr = (StructComplex*)input_items[0];
+    StructComplex* optr = (StructComplex*)output_items[0];
+    float* freq_optr = output_items.size() >= 2 ? (float*)output_items[1] : NULL;
+    float* phase_optr = output_items.size() >= 3 ? (float*)output_items[2] : NULL;
+    float* error_optr = output_items.size() >= 4 ? (float*)output_items[3] : NULL;
+
+#if !defined(__FMA__)
+    float i_r, i_i;
+#endif
+
+    float n_r, n_i, o_r, o_i;
+
+    std::vector<tag_t> tags;
+    get_tags_in_range(tags,
+                      0,
+                      nitems_read(0),
+                      nitems_read(0) + noutput_items,
+                      pmt::intern("phase_est"));
+
+    for (int i = 0; i < noutput_items; i++) {
+        if (!tags.empty()) {
+            if (tags[0].offset - nitems_read(0) == (size_t)i) {
+                d_phase = (float)pmt::to_double(tags[0].value);
+                tags.erase(tags.begin());
+            }
+        }
+
+        if ((d_phase > CL_TWO_PI) || (d_phase < CL_MINUS_TWO_PI)) {
+#if defined(__FMA__)
+            d_phase = __builtin_fmaf(
+                d_phase, CL_ONE_OVER_2PI, -(float)((int)(d_phase * CL_ONE_OVER_2PI)));
+#else
+            d_phase =
+                d_phase * CL_ONE_OVER_2PI - (float)((int)(d_phase * CL_ONE_OVER_2PI));
+#endif
+            d_phase = d_phase * CL_TWO_PI;
+        }
+        n_i = sinf(-d_phase);
+        n_r = cosf(-d_phase);
+
+#if defined(__FMA__)
+        o_r = __builtin_fmaf(iptr[i].real, n_r, -iptr[i].imag * n_i);
+        o_i = __builtin_fmaf(iptr[i].real, n_i, iptr[i].imag * n_r);
+#else
+        i_r = iptr[i].real;
+        i_i = iptr[i].imag;
+        o_r = (i_r * n_r) - (i_i * n_i);
+        o_i = (i_r * n_i) + (i_i * n_r);
+#endif
+        optr[i].real = o_r;
+        optr[i].imag = o_i;
+
+        d_error = (o_r > 0 ? 1.0 : -1.0) * o_i - (o_i > 0 ? 1.0 : -1.0) * o_r;
+        d_error = 0.5 * (std::abs(d_error + 1) - std::abs(d_error - 1));
+
+#if defined(__FMA__)
+        d_freq = __builtin_fmaf(d_beta, d_error, d_freq);
+        d_phase = d_phase + __builtin_fmaf(d_alpha, d_error, d_freq);
+#else
+        d_freq = d_beta * d_error + d_freq;
+        d_phase = d_phase + d_alpha * d_error + d_freq;
+#endif
+
+        if (d_freq > d_max_freq)
+            d_freq = d_max_freq;
+        else if (d_freq < d_min_freq)
+            d_freq = d_min_freq;
+
+        if (freq_optr != NULL)
+            freq_optr[i] = d_freq;
+        if (phase_optr != NULL)
+            phase_optr[i] = d_phase;
+        if (error_optr != NULL)
+            error_optr[i] = d_error;
+    }
+
+    return noutput_items;
+}
+
+int costas_loop_cc_impl::work(int noutput_items,
+                              gr_vector_const_void_star& input_items,
+                              gr_vector_void_star& output_items)
+{
+    return workDefault(noutput_items, input_items, output_items);
+
+    switch (d_order) {
```

Comment:
Because of the `return` just above, none of this code will be executed ... is that the intent? If so, please remove the extra code. If not, please remove the `return` and/or fixup this section to make sense for the code intent.

Reply:
I'm going to go back to the pure cpp test timing code I have in gr-lfast and confirm the real() imag() item.  When I originally wrote it a couple years back I was splitting hairs on performance and timing a line change at a time to squeeze out every last drop of performance and it made a difference.  Would you know off the top of your head if -O2 or -O3 is the default in the GNU Radio build?  I think the value in some of those types of explicit changes comes down to trust (I can be a control freak about it) and whether "the compiler should" versus every compiler "always will" optimize out.  But I'll revalidate it line-by-line just to confirm there's a difference or not.  I think it'll be a good exercise to redo.

Reply:
I found the flag setting in the CMakeCache file.  Looks like the general release build is -O3, and with debug is -O2.